Add utility to get all unsafe globals in checkpoint (no pickletools dependency)

mikaylagawarecki opened the pull request:
Fixes https://github.com/pytorch/pytorch/issues/129698

https://github.com/pytorch/pytorch/pull/139106 without pickletools

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #139303
* #137602
* #139433
* __->__ #139221
* #138936

mikaylagawarecki committed:
Add utility to get all unsafe globals in checkpoint (no pickletools dependency)

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add utility to get all unsafe globals in checkpoint (no pickletools dependency)"

https://github.com/pytorch/pytorch/pull/139106 without pickletools




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add utility to get all unsafe globals in checkpoint (no pickletools dependency)"

https://github.com/pytorch/pytorch/pull/139106 without pickletools




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add utility to get all unsafe globals in checkpoint (no pickletools dependency)"

https://github.com/pytorch/pytorch/pull/139106 without pickletools




[ghstack-poisoned]

malfet commented on torch/_weights_only_unpickler.py:
```diff
@@ -207,6 +207,80 @@ def _get_allowed_globals():
+            raise EOFError
+        assert isinstance(key, bytes_types)
+        if key[0] == GLOBAL[0]:
+            module = readline()[:-1].decode("utf-8")
+            name = readline()[:-1].decode("utf-8")
+            if protocol == 2:
+                if (module, name) in NAME_MAPPING:
+                    module, name = NAME_MAPPING[(module, name)]
+                elif module in IMPORT_MAPPING:
+                    module = IMPORT_MAPPING[module]
```
This code is identical with one define in `def load`, why not refactor it into one function.
Also, just curious, is there a harm for trying to do NAME_MAPPING/IMPORT_MAPPING even if protocol version is different? (It would be a no-op for protocol 3+, isn't it?)

mikaylagawarecki replied:
will refactor!

Yep there is likely no harm trying to do NAME_MAPPING/IMPORT_MAPPING regardless of protocol

malfet commented on torch/_weights_only_unpickler.py:
```diff
@@ -207,6 +207,80 @@ def _get_allowed_globals():
+            if strlen > maxsize:
+                raise UnpicklingError("String is too long")
+            read(strlen)
+        elif key[0] == SHORT_BINSTRING[0]:
+            strlen = read(1)[0]
+            read(strlen)
+        elif key[0] == LONG1[0]:
+            n = read(1)[0]
+            read(n)
```
Nit
```suggestion
        elif key[0] in {SHORT_BINSTRING[0], LONG1[0]}:
            n = read(1)[0]
            read(n)
```

malfet reviewed (approved):
LGTM, imo it's much more explicit what are we doing here, and what are the guarantees...

mikaylagawarecki committed:
Update on "Add utility to get all unsafe globals in checkpoint (no pickletools dependency)"

https://github.com/pytorch/pytorch/pull/139106 without pickletools




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add utility to get all unsafe globals in checkpoint (no pickletools dependency)"

https://github.com/pytorch/pytorch/pull/139106 without pickletools




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add utility to get all unsafe globals in checkpoint (no pickletools dependency)"

https://github.com/pytorch/pytorch/pull/139106 without pickletools




[ghstack-poisoned]

malfet commented on torch/serialization.py:
```diff
@@ -316,6 +317,43 @@ class safe_globals(_weights_only_unpickler._safe_globals):
+    safe_global_strings = set(_weights_only_unpickler._get_allowed_globals().keys())
+
+    with _open_file_like(f, "rb") as opened_file:
+        if _is_zipfile(opened_file):
+            with _open_zipfile_reader(opened_file) as zip_file:
+                if _is_torchscript_zip(zip_file):
+                    raise ValueError(
+                        "Expected input to be a checkpoint returned by torch.save but got a torchscript checkpoint"
+                    )
+                data_file = io.BytesIO(zip_file.get_record("data.pkl"))
+                all_globals = _weights_only_unpickler.get_globals_in_pkl(data_file)
+                return list(all_globals.difference(safe_global_strings))
+        else:
+            raise ValueError("Expected input to be a checkpoint returned by torch.save")
```
Nit (dismantle pyramid of doom by a bit)
```suggestion
        if not _is_zipfile(opened_file):
            raise ValueError("Expected input to be a checkpoint returned by torch.save")
        with _open_zipfile_reader(opened_file) as zip_file:
            if _is_torchscript_zip(zip_file):
                raise ValueError(
                    "Expected input to be a checkpoint returned by torch.save but got a torchscript checkpoint"
                )
            data_file = io.BytesIO(zip_file.get_record("data.pkl"))
            all_globals = _weights_only_unpickler.get_globals_in_pkl(data_file)
            return list(all_globals.difference(safe_global_strings))
```

mikaylagawarecki committed:
Update on "Add utility to get all unsafe globals in checkpoint (no pickletools dependency)"

https://github.com/pytorch/pytorch/pull/139106 without pickletools




[ghstack-poisoned]

mikaylagawarecki commented:
@pytorchbot merge
